### PR TITLE
Show package release dates to make it easier to identify updated packages

### PR DESCRIPTION
**Pull request by harupy**
Signed-off-by: harupy <hkawamura0130@gmail.com>

## What changes are proposed in this pull request?

This PR proposes to show package release dates to make it easier to identify updated packages when we encounter issues:

The output looks like this:

```
Package                              Version             Release Date
---------------------------------------------------------------------
astroid                              2.8.4               2021-10-25  
numpy                                1.21.3              2021-10-20  
xgboost                              1.5.0               2021-10-17  
mleap                                0.18.1              2021-09-27  
platformdirs                         2.4.0               2021-09-25  
pylint                               2.11.1              2021-09-16  
pytest                               6.2.5               2021-08-30  
...
```

## How is this patch tested?

NA

### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: dev/show_package_release_dates.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+import subprocess
+import requests
+from concurrent.futures import ThreadPoolExecutor
+import traceback
```
Comment (harupy) [inline dev/show_package_release_dates.py:4]: This script only contains standard libraries.

File: dev/show_package_release_dates.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    ):
+        print(
+            package.ljust(package_legnth),
+            version.ljust(version_length),
+            release_date.ljust(release_date_length),
+        )
+
+
+if __name__ == "__main__":
+    main()
```
Comment (harupy) [inline dev/show_package_release_dates.py:79]: Running this script takes 7 seconds in the flavors check where we install many ML libraries:

https://github.com/mlflow/mlflow/runs/4066994172?check_suite_focus=true#step:7:728

```
Mon, 01 Nov 2021 11:15:47 GMT ++ python dev/show_package_release_dates.py
Mon, 01 Nov 2021 11:15:54 GMT Package                              Version     Release Date
```

File: dev/show_package_release_dates.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    lines = pip_list_stdout.splitlines()[2:]  # `[2:]` removes the header
+    return [
+        # Extract package and version
+        line.split()[:2]
+        for line in lines
+    ]
+
+
+def get_release_date(package, version):
+    resp = requests.get(f"https://pypi.python.org/pypi/{package}/json")
```
Comment (BenWilson2) [inline dev/show_package_release_dates.py:27]: requests lib has a timeout default of 'None' which will indefinitely hold on to the connection until a server response arrives through the socket. Perhaps add a 10s timeout (or something?) to bomb this out if some packet loss occurs.

File: dev/show_package_release_dates.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    try:
+        return future.result()
+    except Exception:
+        traceback.print_exc()
+        return if_error
+
+
+def main():
+    distributions = get_distributions()
+    with ThreadPoolExecutor() as executor:
```
Comment (BenWilson2) [inline dev/show_package_release_dates.py:53]: Default max_workers value of os.cpu_count() * 5 is a 'goldilock zone' default that strikes a balance between moderate CPU-bound tasks and I/O bound tasks. Py3.8 `concurrent.futures.ThreadPoolExecutor` did a config change for this default that might be helpful in the case of spotty socket connection to pypy, reserving a few threads for longer response times without incurring excessive context switching. 
```python
with ThreadPoolExecutor(max_workers=min(32, os.cpu_count() + 4)):
```
Just a thought.

File: dev/show_package_release_dates.py (hunk 3 above)
Comment (harupy) [reply dev/show_package_release_dates.py:27]: Sounds good, thanks for pointing this out!

File: dev/show_package_release_dates.py (hunk 4 above)
Comment (harupy) [reply dev/show_package_release_dates.py:53]: Thanks for the knowledge! I just read https://bugs.python.org/issue35279. Using `min(32, os.cpu_count() + 4)` sounds good to me.

**Review by BenWilson2 — APPROVED** [review]
Looks great! And the speed improvement for lint checks is very much appreciated! :)

**Pull request merged**